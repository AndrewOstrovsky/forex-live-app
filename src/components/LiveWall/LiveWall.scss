.c-form {
    overflow: hidden;
    display: flex;
    align-items: center;

    &-error {
        margin-left: 10px;
        padding: 10px;
        color: #D8000C;
        background-color: #FFBABA;
    }
}