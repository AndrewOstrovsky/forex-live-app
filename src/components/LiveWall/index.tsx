import React, { ChangeEvent, FormEvent, useState } from 'react';

import searchData from '../../helpers/searchData'
import Table from '../Table'
import FormInput from '../FormInput'
import Button from '../Button'
import LoadingSpinner from '../LoadingSpinner'

import './LiveWall.scss'
import { IDataItem } from '../../helpers/searchData/types';
import { REQUEST_INTERVAL } from '../../constants'

type TTimer = undefined | number

let timer: TTimer

const LiveWall = () => {
  const [data, setData] = useState([])
  const [isLoading, setIsLoading] = useState(false)
  const [error, setError] = useState(null)
  const [query, setQuery] = useState('')

  const onError = (message: string) => {
    setData([])
    setError(message)
    setIsLoading(false)
    if (timer) {
      clearInterval(timer)
    }
  }

  const onSucess = (data: IDataItem[]) => {
    setData(data)
    setIsLoading(false)
  }

  const onSubmit = (event: FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    setIsLoading(true)
    setError(null)

    searchData(query, onSucess, onError)

    if (timer) {
      clearInterval(timer)
    }
    timer = window.setInterval(() => searchData(query, onSucess, onError), REQUEST_INTERVAL)
  }

  const onChange = (event: ChangeEvent<HTMLInputElement>) => setQuery(event.target.value)

  return(
        <React.Fragment>
            <form className="c-form" onSubmit={onSubmit}>
                <FormInput onChange={onChange} />
                <Button>Search</Button>
                { isLoading && <LoadingSpinner />}
                { error && <span className="c-form-error">{error}</span> }
            </form>
            <Table data={data}/>
        </React.Fragment>
  );
}

export default LiveWall