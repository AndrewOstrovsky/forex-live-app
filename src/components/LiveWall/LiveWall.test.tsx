import React from 'react';

import renderer from 'react-test-renderer';
import LiveWall from '.';

it('renders correctly', () => {
  const component = renderer
    .create(<LiveWall />)
    .toJSON();
  expect(component).toMatchSnapshot();
});