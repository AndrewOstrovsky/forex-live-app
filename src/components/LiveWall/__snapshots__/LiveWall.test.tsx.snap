// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <form
    className="c-form"
    onSubmit={[Function]}
  >
    <input
      className="c-form-input"
      onChange={[Function]}
      type="text"
    />
    <button
      className="c-button"
      type="submit"
    >
      Search
    </button>
  </form>,
  <table
    className="c-table"
  >
    <thead>
      <tr
        className="c-table-row-heading"
      >
        <th
          className="c-table-cell"
        >
          Pair
        </th>
        <th
          className="c-table-cell"
        >
          Bid
        </th>
        <th
          className="c-table-cell"
        >
          Ask
        </th>
        <th
          className="c-table-cell"
        >
          Change %
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;
